client/asset/dcr/externaltx.go

```diff
@@ -0,0 +1,444 @@
+// mainchain, this method matches the output's pkScript against multiple block
+// cfilters (starting from the block containing the output up till the current
+// best block) in an attempt to find the block containing the output spender.
+// If found, the block hash and height is recorded.
+func (dcr *ExchangeWallet) findExternalTxOutputSpender(output *externalTxOutput, txBlockHash *chainhash.Hash) (bool, error) {
+	// If the spender of this tx output is not yet known, and some other
+	// process got here first, a search for the tx output spender might
+	// be underway already. Wait until any previous search completes and
+	// releases this lock.
+	output.spender.scanMtx.Lock()
+	defer output.spender.scanMtx.Unlock()
+
+	// Check if the spender of this tx output was found in a previous search
+	// attempt.
+	spenderBlockHash, err := output.spender.relevantBlockHash(dcr.getBlockHash)
+	if err != nil {
+		return false, err
+	}
+	if spenderBlockHash != nil {
+		return true, nil
+	}
+
+	// This tx output is not known to be spent as of last search (if any).
+	// Scan blocks from the lastScannedBlock (if there was a previous scan)
+	// or from the tx block to attempt finding the spender of this output.
+	// Use mainChainAncestor to ensure that scanning starts from a mainchain
+	// block in the event that either tx block or lastScannedBlock have been
+	// re-orged out of the mainchain.
+	var startBlock *chainjson.GetBlockVerboseResult
+	if output.spender.lastScannedBlock == nil {
+		_, startBlock, err = dcr.mainChainAncestor(txBlockHash)
+	} else {
+		_, startBlock, err = dcr.mainChainAncestor(output.spender.lastScannedBlock)
+	}
+	if err != nil {
+		return false, err
+	}
+
+	// Search for this output's spender in the blocks between startBlock and bestBlock.
+	bestBlock := dcr.cachedBestBlock()
+	dcr.log.Debugf("Searching for the tx that spends output %s in blocks %d (%s) to %d (%s).",
+		output, startBlock.Height, startBlock.Hash, bestBlock.height, bestBlock.hash)
+	for blockHeight := startBlock.Height; blockHeight <= bestBlock.height; blockHeight++ {
+		blockHash, err := dcr.node.GetBlockHash(dcr.ctx, blockHeight)
+		if err != nil {
+			return false, translateRPCCancelErr(err)
+		}
+		blockFilter, err := dcr.getBlockFilterV2(blockHash)
+		if err != nil {
+			return false, err
+		}
+		if !blockFilter.Match(output.PkScript) {
+			output.spender.lastScannedBlock = blockHash
+			continue // check next block's filters (blockHeight++)
+		}
+		dcr.log.Debugf("Block %d (%s) likely contains a tx that spends %s. Confirming.",
+			blockHeight, blockHash, output)
+		blk, err := dcr.getBlock(blockHash, true)
+		if err != nil {
+			return false, err
+		}
+		for i := range blk.RawTx {
+			blkTx := &blk.RawTx[i]
+			if txSpendsOutput(blkTx, output) {
+				dcr.log.Debugf("Found spender for output %s in block %d (%s), spender tx hash %s.",
+					output, blockHeight, blockHash, blkTx.Txid)
+				output.spender.relevantBlock = &block{hash: blockHash, height: blockHeight}
```

Comment:
It's a little curious to me that `spender` does not actually record the spending txn, just the block that it's in.  Unlike the redemption search, I suppose we only care to know **if it is spent** for (1) audit contract, and (2) `(*ExchangeWallet).Confirmations` method (for `shouldBeginFindRedemption` and `counterPartyConfirms`).  Is that right?

Reply:
Right. I'd thought to record the actual spending tx but it doesn't really have any use.